tests/scripts/thread-cert/test_ipv6_source_selection.py

```diff
@@ -0,0 +1,147 @@
+        multicast_addr = 'ff03::1'
+        external_addr = '2007::1'
+
+        router_rloc = self.nodes[ROUTER].get_rloc()
+        router_linklocal = self.nodes[ROUTER].get_linklocal()
+        router_mleid = self.nodes[ROUTER].get_mleid()
+
+        # source check - rloc source for rloc destination
+        self.assertTrue(self.nodes[ROUTER].ping(leader_rloc))
+        router_msgs = self.simulator.get_messages_sent_by(ROUTER)
+        msg = router_msgs.get_icmp_message(ipv6.ICMP_ECHO_REQUEST)
+        msg.assertSentFromSourceAddress(router_rloc)
+
+        # source check - mleid source for aloc destination
+        self.assertTrue(self.nodes[ROUTER].ping(leader_aloc))
+        router_msgs = self.simulator.get_messages_sent_by(ROUTER)
+        msg = router_msgs.get_icmp_message(ipv6.ICMP_ECHO_REQUEST)
+        msg.assertSentFromSourceAddress(router_mleid)
+
+        # source check - mleid source for mleid destination
+        self.assertTrue(self.nodes[ROUTER].ping(leader_mleid))
+        router_msgs = self.simulator.get_messages_sent_by(ROUTER)
+        msg = router_msgs.get_icmp_message(ipv6.ICMP_ECHO_REQUEST)
+        msg.assertSentFromSourceAddress(router_mleid)
+
+        # source check - link local source source for link local address
+        self.assertTrue(self.nodes[ROUTER].ping(leader_linklocal))
+        router_msgs = self.simulator.get_messages_sent_by(ROUTER)
+        msg = router_msgs.get_icmp_message(ipv6.ICMP_ECHO_REQUEST)
+        msg.assertSentFromSourceAddress(router_linklocal)
+
+        # source check - mleid source for realmlocal multicast destination
+        self.assertTrue(self.nodes[ROUTER].ping(multicast_addr))
+        router_msgs = self.simulator.get_messages_sent_by(ROUTER)
+        msg = router_msgs.get_icmp_message(ipv6.ICMP_ECHO_REQUEST)
+        msg.assertSentFromSourceAddress(router_mleid)
+
+        # GUA and default gateway
+        self.nodes[LEADER].add_prefix('2001::/64', 'paros')
+        self.nodes[LEADER].register_netdata()
+        self.simulator.go(5)
+
+        # Set lowpan context of sniffer
+        self.simulator.set_lowpan_context(1, '2001::/64')
+
+        # flushes message queue before next ping
+        router_msgs = self.simulator.get_messages_sent_by(ROUTER)
+
+        # source check - GUA source for GUA destination
+        leader_gua = self.nodes[LEADER].get_addr("2001::/64")
+        router_gua = self.nodes[ROUTER].get_addr("2001::/64")
+        print(leader_gua)
+        print(router_gua)
+        self.assertTrue(leader_gua is not None)
+        self.assertTrue(router_gua is not None)
+        self.assertTrue(self.nodes[ROUTER].ping(leader_gua))
+        router_msgs = self.simulator.get_messages_sent_by(ROUTER)
+        msg = router_msgs.get_icmp_message(ipv6.ICMP_ECHO_REQUEST)
+        msg.assertSentFromSourceAddress(router_gua)
+
+        # source check - GUA source for external address (default route)
+        self.nodes[ROUTER].ping(external_addr)
```

Comment:
not: add assert as above

Reply:
Different from the other pings, the ping here doesn't has the reply.